apps/performance-tests/src/unified-selection/SelectionScope.test.ts

```diff
@@ -0,0 +1,166 @@
+    elementIds: createElementKeys({ start: 16, count: 100_000 }),
+    scope: "functional",
+    expectedCounts: [{ className: "Functional.FunctionalComposite", count: 50_000 }],
+  });
+
+  runSelectionScopeTest({
+    testName: "parent functional selection for 50k 3D elements",
+    iModelName: "50k nested functional 3D elements",
+    elementIds: createElementKeys({ start: 16, count: 100_000 }),
+    scope: { id: "functional", ancestorLevel: 1 },
+    expectedCounts: [{ className: "Functional.FunctionalComposite", count: 49_000 }],
+  });
+
+  runSelectionScopeTest({
+    testName: "top ancestor functional selection for 50k 3D elements",
+    iModelName: "50k nested functional 3D elements",
+    elementIds: createElementKeys({ start: 16, count: 100_000 }),
+    scope: { id: "functional", ancestorLevel: -1 },
+    expectedCounts: [{ className: "Functional.FunctionalComposite", count: 1_000 }],
+  });
+
+  runSelectionScopeTest({
+    testName: "functional selection for 50k 2D elements",
+    iModelName: "10k nested functional 2D elements",
+    elementIds: createElementKeys({ start: 17, count: 20_000 }),
+    scope: "functional",
+    expectedCounts: [{ className: "Functional.FunctionalComposite", count: 10_000 }],
+  });
+
+  runSelectionScopeTest({
+    testName: "parent functional selection for 50k 2D elements",
+    iModelName: "10k nested functional 2D elements",
+    elementIds: createElementKeys({ start: 17, count: 20_000 }),
+    scope: { id: "functional", ancestorLevel: 1 },
+    expectedCounts: [{ className: "Functional.FunctionalComposite", count: 9_000 }],
+  });
+
+  runSelectionScopeTest({
+    testName: "top ancestor functional selection for 50k 2D elements",
+    iModelName: "10k nested functional 2D elements",
+    elementIds: createElementKeys({ start: 17, count: 20_000 }),
+    scope: { id: "functional", ancestorLevel: -1 },
+    expectedCounts: [{ className: "Functional.FunctionalComposite", count: 1_000 }],
+  });
+});
+
+function runSelectionScopeTest(
+  testProps: {
+    iModelName: IModelName;
+    elementIds: string[];
+    scope: ComputeSelectionProps["scope"];
+    expectedCounts?: { className: string; count: number }[];
+  } & Omit<RunOptions<never>, "setup" | "test" | "cleanup">,
+) {
+  const { iModelName } = testProps;
+  run({
+    ...testProps,
+    setup: () => {
+      const iModel = SnapshotDb.openFile(Datasets.getIModelPath(iModelName));
+      const queryExecutor = createECSqlQueryExecutor(iModel);
+
+      return {
+        iModel,
+        queryExecutor,
+      };
+    },
+    test: async (props) => {
+      const iterator = computeSelection({ elementIds: testProps.elementIds, scope: testProps.scope, queryExecutor: props.queryExecutor });
+      const counts = new Map<string, number>();
+
+      for await (const instanceKey of iterator) {
+        const count = counts.get(instanceKey.className) ?? 0;
+        counts.set(instanceKey.className, count + 1);
+      }
```

Comment:
I'm a worried the Map lookup is causing the tests to be slower than necessary. Can you try the following to see if there's any positive effect?

```suggestion
      const counts = new Map<string, { count: number }>();

      for await (const instanceKey of iterator) {
        const res = counts.get(instanceKey.className);
        if (res) {
          ++res.count;
        } else {
          res.set(instanceKey.className, { count: 1 });
        }
      }
```

Reply:
Switched to the suggestion, does not seem to have any real effect on the results.